vassal-app/src/main/java/VASSAL/build/module/index/IndexManager.java

```diff
@@ -0,0 +1,179 @@
+ * at http://www.opensource.org.
+ */
+package VASSAL.build.module.index;
+
+import VASSAL.build.GameModule;
+import VASSAL.build.module.Chatter;
+import VASSAL.build.module.Map;
+import VASSAL.counters.GamePiece;
+import VASSAL.counters.Stack;
+
+import java.awt.Point;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * The IndexManager maintains a set of indexes on GamePieces that will primarily be used by
+ * the GKC Fast-match mechanism to quickly find sets of pieces that meet certain critera.
+ *
+ * It maintains the following indexes:
+ *  o A per-map quadtree of piece locations
+ *  o A per-map cross-reference of values of the Currentzone and LocationName properties
+ *
+ *  The indexes for each Map are stored in a VassalMapPieceIndex object
+ *
+ *  NOTE: Pieces not on Maps are NOT included in any index.
+ */
+public class IndexManager {
+
+  /**
+   * HashMap of the indexes created so far
+   */
+  private final java.util.Map<Map, VassalMapPieceIndex> indexes = new HashMap<>();
+
+  /**
+   * Return the combined index object for the specified Map, create one if needed
+   *
+   * @param map Map
+   * @return Index
+   */
+  private VassalMapPieceIndex getIndex(Map map) {
+    VassalMapPieceIndex index = indexes.get(map);
+    if (index == null) {
+      index = new VassalMapPieceIndex(map);
+      indexes.put(map, index);
+    }
+    return index;
+  }
+
+  /**
+   * Clear and rebuild all indexes using pieces currently existing on boards.
+   * Will usually only be called at the start of a game session after initial load of gamestate is complete.
+   */
+  public void rebuild() {
+    new Chatter.DisplayText(GameModule.getGameModule().getChatter(), "Indexing pieces...").execute();
+    clearAll();
+    for (final GamePiece piece : GameModule.getGameModule().getGameState().getAllPieces()) {
+      final Map map = piece.getMap();
+      if (map != null) {
+        pieceMoved(piece, map);
+      }
+    }
+  }
+
+  /**
+   * Clear all indexes
+   */
+  public void clearAll() {
+    indexes.clear();
+  }
+
+  /**
+   * A piece has just moved or been placed on the board, update any location based indexes
+   * and create any property based indexes if we have not seen this piece before
+   *
+   * @param piece Piece that moved
+   * @param map   Map that the piece has arrived on
+   */
+  public void pieceMoved(GamePiece piece, Map map) {
+    if (map != null) {
+      if (piece instanceof Stack) {
+        for (final GamePiece p : ((Stack) piece).asList()) {
+          getIndex(map).addOrUpdatePiece(p);
+        }
+      }
+      else {
+        getIndex(map).addOrUpdatePiece(piece);
+      }
+    }
+  }
+
+  /**
+   * A piece has been removed from a Map
+   *
+   * @param piece Removed piece
+   * @param map   Map piece was removed from
+   */
+  public void pieceRemoved(GamePiece piece, Map map) {
+    if (map != null) {
+      if (!(piece instanceof Stack)) {
+        if (piece instanceof Stack) {
```

Comment:
This can't be right. You've checked `!(piece instanceof Stack)` and then `piece instanceof Stack`.

Reply:
One too many ifs